README.md
# ETL for CoDa

This repository contains the ETL (extract, transform, and load) pipeline for the Cooperation Databank, which is hosted on the CoDa Triply instance at [data.cooperationdatabank.org](https://data.cooperationdatabank.org).

### Running individual scripts locally

- Make sure the directory `./data/input` contains all the data needed by the script you want to run
- Include required environment variables before running the script, e.g.
```
source .paths && python3 ./convert-indicators.py
```

### Data

The data from the ETL is stored in the `output_graphs/` directory. The files that are converted have the extension `.n3` and can be uploaded as linked data, on the instance of preference, or exposed as downloadable content on the website.


CHANGELOG.md
# CoDa ETL changelog

## Version 1.6.21 2021-03-10

* Fixed computation of logit rates of coopreation for withdrawals and computation of coefficient of variance
* Fixed computation of variance for contributions

## Version 1.6.17 2020-11-25

* Public launch


src/EffectSizeFunctions.R
###############################################################################
# Comparing two independent samples ###########################################
###############################################################################

# Effect size for comparison of two independent groups ########################
# Computed from means, standard deviations, and sample sizes
es_from_means = function(data, effsize){
  es_from_means <- data %>%
    filter(`BS/WS` == 1 &
             !(SD.t1 == 0 | SD.t2 == 0) &
             (`Main Effect / Interaction` == 1 |
                `Main Effect / Interaction` == 3)) %>%
    dplyr::select(effect_ID, 
                  M.t1, M.t2, SD.t1, SD.t2, n.t1, n.t2) %>%
    drop_na() %>%
    effect_sizes(grp1m = M.t1, grp1sd = SD.t1, grp1n = n.t1, 
                 grp2m = M.t2, grp2sd = SD.t2, grp2n = n.t2,
                 study = effect_ID, 
                 fun = "esc_mean_sd",
                 es.type = effsize)
}

# Effect size for comparison of two independent groups ########################
# Computed from t-statistics and sample sizes
# If Direction == 1, the t value should be negative; 
# if Direction == 0, the t value should be positive; 
# otherwise invert the t statistic.
es_from_tn = function(data, effsize){
  es_from_means <- data %>%
    filter(`BS/WS` == 1 &
             Test == "t" &
             `# Levels` == 2 &
             (`Main Effect / Interaction` == 1 |
                `Main Effect / Interaction` == 3)) %>%
    dplyr::select(effect_ID, 
                  Statistics, n.t1, n.t2, Direction) %>%
    mutate(Statistics = case_when(Direction == 1 & Statistics > 0 ~ -Statistics,
                                  Direction == 0 & Statistics < 0 ~ Statistics * (-1))) %>%
    drop_na() %>%
    effect_sizes(grp1n = n.t1, grp2n = n.t2, t = Statistics,
                 study = effect_ID, 
                 fun = "esc_t",
                 es.type = effsize)
  return(es_from_means) }

# Cohen's d for two independent groups ########################################
# Computed from t-statistics and degrees of freedom
es_from_tdf = function(data, effsize){
  es_from_means <- data %>%
    mutate(df1 = df1 + 1) %>%
    filter(`BS/WS` == 1 &
             Test == "t" &
             `# Levels` == 2 &
             (`Main Effect / Interaction` == 1 |
                `Main Effect / Interaction` == 3)) %>%
    dplyr::select(effect_ID, 
                  Statistics, df1) %>%
    drop_na() %>%
    effect_sizes(totaln = df1, t = Statistics,
                 study = effect_ID, 
                 fun = "esc_t",
                 es.type = effsize)
  return(es_from_means) }

# Effect size for comparison of two independent groups ########################
# Computed from proportions and sample sizes
es_from_prop = function(data, effsize){
  es_from_prop <- data %>%
    filter(`BS/WS` == 1 &
             !(is.na(`P(C).t1`) | is.na(`P(C).t2`)) &
             (`Main Effect / Interaction` == 1 |
                `Main Effect / Interaction` == 3)) %>%
    dplyr::select(effect_ID, 
                  `P(C).t1`, `P(C).t2`, n.t1, n.t2) %>%
    drop_na() %>%
    effect_sizes(prop1event = `P(C).t1`, grp1n = n.t1, 
                 prop2event = `P(C).t2`, grp2n = n.t2,
                 study = effect_ID, 
                 fun = "esc_bin_prop",
                 es.type = effsize)
}

###############################################################################
# Comparing two dependence samples ############################################
###############################################################################

es_from_means_dep <- function(data, effsize){
  es_from_means_dep <- data %>%
    filter(`BS/WS` == 2 &
             !(SD.t1 == 0 | SD.t2 == 0) &
             (`Main Effect / Interaction` == 1 |
                `Main Effect / Interaction` == 3) &
             n.t1 == n.t2) %>%
    dplyr::select(effect_ID, 
                  M.t1, M.t2, SD.t1, SD.t2, n.t1, n.t2) %>%
    drop_na() %>%
    mutate(r = 0.5,
           study = effect_ID,
           es = (M.t1-M.t2) / sqrt(SD.t1^2 + SD.t2^2 - 2 * r * SD.t1 * SD.t2),
           weight = NA,
           sample.size = n.t1,
           se = NA,
           var = (1/n.t1) + (es^2 / (2*n.t1)),
           ci.lo = es - 1.96*(sqrt(var)/sqrt(sample.size)),
           ci.hi = es + 1.96*(sqrt(var)/sqrt(sample.size)),
           measure = effsize) %>%
    select(study:measure)
}

###############################################################################
# Correlations ################################################################
###############################################################################

# Effect size for association of two continuous variables
# Computed from correlation coefficient and sample sizes
es_from_rn = function(data, effsize){
  es_from_rn <- data %>%
    filter(`Effect size measure` == "r") %>%
    mutate(n.t1 = ifelse(is.na(n.t1), N, n.t1)) %>%
    select(effect_ID, ES, n.t1) %>%
    drop_na() %>%
    mutate(var_r = (1 - ES^2)/(n.t1 - 2),
           measure = effsize) %>%
    mutate(study = effect_ID,
           es = ifelse(measure == "r", ES,
                       ifelse(measure == "d", (2*ES)/sqrt(1-ES^2), NA)),
           weight = NA,
           sample.size = n.t1,
           var = ifelse(measure == "r", var_r,
                        ifelse(measure == "d", (4*var_r)/(1-ES^2)^3, NA)),
           se = NA,
           ci.lo = es - 1.96 * sqrt(var/sample.size),
           ci.hi = es + 1.96 * sqrt(var/sample.size)) %>%
    select(study, es, weight, sample.size, se, var, ci.lo, ci.hi, measure)
}

src/convert-indicators.py
from rdflib import Graph, Namespace, URIRef, Literal, BNode
from rdflib.namespace import RDF, RDFS, XSD, OWL
import pandas as pd
import os,sys,string
import requests

coda_vocab = Namespace("https://data.cooperationdatabank.org/vocab/")
coda_res = Namespace("https://data.cooperationdatabank.org/id/")
resourceDir = os.getenv('TRIPLY__PATHS__DATA_DIR', ".")


"""
utility fcts
"""
def isEmpty(val):
    emptyVals =[ "", "NA", "NaN" , "N/A" , 'nan' , 'None' , '999' ,'999.0',  'missing' , 'other' , 'others', 'Other', 'Others']
    if str(val) in emptyVals : return True
    return False



def parse(s):
    clean = ""
    for char in s:
        if char in string.punctuation or char == " " :
            clean = clean+'-'
        else :
            clean = clean+char
    return clean


def readCSV(inFile):
    """
    returns csv without headers
    """
    df  = pd.read_csv(inFile,  na_filter = False, header=0, encoding="latin-1" )
    return df


def save_graph(graph, output_name):
    graph.serialize(destination="%s" % output_name, format='trig')


def add_wikidata_info(countries,indicators):
    wikidataQ = """ 
    SELECT DISTINCT * WHERE 
        { ?s ?p ?o. 
          ?s wdt:P984 ?name.
          VALUES ?name { ::slist:: } 
          VALUES ?p { ::plist:: }. 
      }
    """
    wikidataEP = "https://query.wikidata.org/sparql"
    
    wprops = " ".join(i for i in indicators)
    val_countries = " ".join('"%s"' % j for j in countries)
    
    query  = wikidataQ.replace("::slist::",val_countries).replace("::plist::", wprops)
    
    try:
        resp = requests.get(wikidataEP, params={'query': query, 'format': 'json'}).json()
    except requests.exceptions.RequestException as e:
        print("RequestException for Wikidata %s" % str(e) )
        resp = []
    
    return resp

def add_fb_info(countries, indicators):
    fbQuery = """
    PREFIX wdt: <http://www.wikidata.org/prop/direct/>
    SELECT ?valueName ?p ?o ?name WHERE {
    service <https://query.wikidata.org/sparql> {
        ?s wdt:P984 ?name.
    	?s wdt:P297 ?value.
          BIND (IRI(concat("http://www.daml.org/2001/09/countries/fips#",?value) ) as ?valueName)
        } 
    	?valueName ?p ?o    
      values ?p { ::flist:: }
      values ?name { ::nlist:: }
    }
    
    """
    
    fbEndpoint = "https://api.triplydb.com/datasets/rvo/landen-portaal/services/landen-portaal/sparql"
    
    fblist = " ".join("<%s>" % i for i in indicators)
    val_countries = " ".join('"%s"' % j for j in countries)
    
    query = fbQuery.replace("::flist::", fblist).replace("::nlist::", val_countries)
    
    try:
        resp = requests.get(fbEndpoint, params={'query': query, 'format': 'json' }).json() ## canot make , 'Content-Type':'application/json' work
    except requests.exceptions.RequestException as e:
        print("RequestException for Factbook %s" % str(e) )
        resp = []
    
    return resp

if __name__ == "__main__":
    print('######\n Converting country indicators ...')
    
    vocab = readCSV(resourceDir + "/input/indicators-vocab.csv")

    indicators = list()
    wd_indicators = list()
    fb_indicators = list()
    
    v = Graph(identifier=Namespace("https://data.cooperationdatabank.org/countryVocab"))
    v.namespace_manager.bind("cd", Namespace("https://data.cooperationdatabank.org/"))
    v.namespace_manager.bind("dct", Namespace("http://purl.org/dc/terms/"))
    
    for ix, construct in vocab.iterrows():
        if 'N' in construct['keep (Y/N)']: continue

        indicator = URIRef(construct['p'])
        v.add((indicator ,RDF.type, RDF.Property))
        if construct['new label'] != '':
            v.add(( indicator , RDFS.label, Literal(construct['new label'],datatype=XSD.string) ))
        elif construct['pAltName'] != '':
            v.add(( indicator , RDFS.label, Literal(construct['pAltName'],datatype=XSD.string) ))
        else :
            v.add(( indicator , RDFS.label, Literal(construct['pLabel'],datatype=XSD.string) ))
        if construct['definition'] != '':
            v.add(( indicator, URIRef("http://purl.org/dc/terms/description"), Literal(construct['definition'], datatype=XSD.string ) ))
        if construct['p'] not in indicators :
            
            ## add to wikidata list
            if construct['gLabel'] == 'wikidata' :
                wd_indicators.append("wdt:%s" % construct['pLabel'])
            if construct['gLabel'] == 'factbook' :
                fb_indicators.append("%s" % construct['p'])
                
            indicators.append(construct['pLabel'])

    print("Country vocab graph has %d triples" % len(v))
    save_graph(v,resourceDir+ "/output_graphs/indicator-vocab.n3")
    
    #### now onto instances
    
    g = Graph(identifier=Namespace("https://data.cooperationdatabank.org/countryIndicators"))
    g.namespace_manager.bind("cd", Namespace("https://data.cooperationdatabank.org/"))
    g.namespace_manager.bind("wdt", Namespace("http://www.wikidata.org/prop/direct/"))
    g.namespace_manager.bind("cdc", Namespace("https://data.cooperationdatabank.org/vocab/"))
    g.namespace_manager.bind("cdp", Namespace("https://data.cooperationdatabank.org/vocab/prop/"))
    g.namespace_manager.bind("fb", Namespace("http://www.daml.org/2001/12/factbook/factbook-ont#"))
    
    country_data = readCSV(resourceDir + "/input/indicators_reversed_all_zscored_coda_2.1.csv")
    
    seen = list()
    
    for ix,country_row in country_data.iterrows() :
        
        country_res =  URIRef(coda_res+"country/"+country_row['code_coda'])

        # this should speed up
        if country_row['code_coda'] not in seen :
            seen.append(country_row['code_coda'])
            
        for indicator,value in country_row['societal_cynicism_zscore':'ESS_avg_importance_loyalty_reversed_zscore'].iteritems():
            if indicator not in indicators : 
                continue
            if not isEmpty(value):

                # storing observations as a BNode for the moment (TODO : qbes )
                # indicators are qb:MeasureProperty
                observation = BNode()

                g.add((observation, RDF.type, coda_vocab.Observation))
                g.add((observation, URIRef(coda_vocab+"prop/value"), Literal(value, datatype=XSD.double ) ))
                g.add((observation, URIRef(coda_vocab+"prop/year"), Literal(country_row['year'], datatype=XSD.gYear ) ))
                g.add((country_res, URIRef(coda_vocab+"prop/"+indicator), observation ))

    
    ### wikidata from https://query.wikidata.org/
    wikiresp = add_wikidata_info(seen,wd_indicators)
    
    for item in wikiresp['results']['bindings']:
        g.add((URIRef("https://data.cooperationdatabank.org/id/country/%s" % item['name']['value']),  URIRef(item['p']['value']), URIRef(item['o']['value']) ))
    
    
    ### factbook from https://triplydb.com/rvo/landen-portaal/sparql/landen-portaal
    factbookresp = add_fb_info(seen,fb_indicators)
 
    for item in factbookresp:  ## NB triply's response is parsed already
       # print(item)
        objectValue = ""  
        if item['o'].startswith('http') : 
            objectValue = URIRef(item['o']) 
        else :
            objectValue = Literal(item['o'])
        g.add((URIRef("https://data.cooperationdatabank.org/id/country/%s" % item['name']), URIRef(item['p']), objectValue ))
    
    print("Country graph has %d triples" % len(g))
    
    save_graph(g,"./graph_temp.n3")
    os.system("sed 's/\-[0-9][0-9]\-[0-9][0-9]\"\^\^xsd:gYear/\"\^\^xsd:gYear/g' graph_temp.n3 >| "+resourceDir+"/output_graphs/country-indicators.n3   ")
    os.remove("graph_temp.n3")


src/EffectSizeComputation.R
###############################################################################
# CODA: DATA WRANGLING AND EFFECT SIZE COMPUTATION ############################
# SIMON COLUMBUS | 2020-06-26 #################################################
###############################################################################


for (p in c("data.table", "esc", "tidyverse")) {
  if (!require(p, character.only = T)) {
    install.packages(p)
    require(p,character.only=TRUE)
  }
}



# LOAD PACKAGES ###############################################################
library(data.table)
library(esc)
library(tidyverse)

#local path
path = Sys.getenv("TRIPLY__PATHS__DATA_DIR")
srcDir = Sys.getenv("TRIPLY_ETL_SRC")
# load ES functions
source(paste0(srcDir,"/EffectSizeFunctions.R"))

# LOAD DATA ###################################################################
keepingTrack <- read.csv(paste0(path,"/input/Keeping_Track.csv"), # Keeping track
                         stringsAsFactors = FALSE) %>%
  # Remove un-checked studies
  filter(!is.na(study_ID) & grepl("4", Re.Checks))
studyCharacteristics <- read.csv(paste0(path,"/input/Study_characteristics.csv"),
                                 stringsAsFactors = FALSE, # Study characteristics
                                 check.names=FALSE,
                                 colClasses=c(rep(NA, 54), rep("NULL", 71),
                                              rep(NA, 2), "NULL", NA, rep("NULL", 26))) %>% # Drop unused columns
  filter(!is.na(study_ID) & !(study_ID == FALSE)) %>%
  filter(`Dilemma type(s)` %in% 1:3) %>%
  distinct(., study_ID, .keep_all = TRUE) # Remove duplicates
treatments <- read.csv(paste0(path,"/input/Treatments.csv"), # Treatments
                       stringsAsFactors = FALSE,
                       check.names=FALSE) %>%
  filter(!is.na(treatment_ID) & treatment_ID != "") %>%
  # Filter out treatments from unchecked studies
  filter(study_ID %in% keepingTrack$study_ID) %>%
  mutate(`n f` = ifelse(!is.na(Remove) & Remove == 1, -1, `n f`)) %>%
  mutate(`n f` = replace_na(`n f`, 0)) %>%
  filter(!(`n f` == -1 | `n f` == -2))  # Remove deleted treatments
ontology <- read.csv(paste0(path,"/input/Ontology.csv"),
                     stringsAsFactors = FALSE, skip = 1,
                     check.names=FALSE) %>% # Ontology annotation
  filter(!is.na(treatment_ID)) %>%               # Removes notes at the bottom
  distinct(., study_ID, variable_ID, treatment_ID, # Remove duplicates
           .keep_all = TRUE)
# MERGE DATA ##################################################################

data <- treatments %>%
  merge(., studyCharacteristics,
        by = c("study_ID"),
        suffixes = c("", "_char")) %>%
  merge(., ontology,
        by = c("study_ID", "variable_ID", "treatment_ID"),
        suffixes = c("", "_ont")) %>%
  #dplyr::select(., !ends_with("_char") & !ends_with("_ont")) %>% # Remove duplicate columns
  select(study_ID:treatment_ID,
         Published_ES:`Study #`,
         `Overlaps with...`:`N.obs/N`,
         IV_description:SD,
         Agent_Role:OtherCoding)

# DATA CLEANING ###############################################################

# Clean up data for computing effect sizes
# This is mostly converting to numeric vectors and removing invalid entries.
dataClean <- data %>%
  mutate_at(vars(`Overall P(C)`:`P(E) contributed`,
                 `BS/WS`, df1:p, ES,
                 n:SD),
            funs(as.numeric(as.character(.)))) %>%
  mutate(p = ifelse(p == 0 | p == 1 | p > 1, "N/A", p)) %>% # Remove invalid p-values
  mutate_if(is.logical, as.character) %>% # mutate logical columns ()
  mutate_if(is.character, funs(recode(na_if(.,""), `N/A` = "-999", MISSING = "-999", `999` = "-999", .missing = "-999"))) %>%
  mutate_if(is.integer, funs(recode(., `999` = -999L))) %>%
  mutate_if(is.double, funs(recode(., `999` = -999.0))) %>%
  mutate_all(funs(replace(., is.na(.), -999))) %>%
  select(-`If other Recr_Meth: Specify`, # Drop unused columns
         -`Iterated # per block:`,
         -`Block #:`,
         -Feedback,
         -IV_description, -DV, -Level, -Agent_Role)

# Compute additional treatment-level variables:
# Proportion of endowment contributed, log-transformed proportion of endowment
# contributed/rate of cooperation and coefficient of variation.

dataClean <- dataClean %>%
  # For single-treatment variables with missing M/P(C)/n, replace the
  # treatment-level value with the study-level value (unless they are
  # coded as a subset in Main Effect / Interaction).
  mutate(M = case_when(M != -999 ~ M,
                       M == -999 & `# Levels` == -999 & !(`Main Effect / Interaction` %in% c(3, 4)) & DV_behavior == "Contribution" ~ 
                         `Overall M cooperation`,
                       M == -999 & `# Levels` == -999 & !(`Main Effect / Interaction` %in% c(3, 4)) & DV_behavior == "Withdrawals" ~ 
                         `Overall M withdrawal`,
                       TRUE ~ -999),
         `P(C)` = case_when(`P(C)` != -999 ~ `P(C)`,
                            `P(C)` == -999 & `# Levels` == -999 & !(`Main Effect / Interaction` %in% c(3, 4)) & DV_behavior == "Cooperation" ~ 
                              `Overall P(C)`,
                            TRUE ~ -999),
         n = case_when(n != -999 ~ n,
                       n == -999 & `# Levels` == -999 & !(`Main Effect / Interaction` %in% c(3, 4)) ~ `N.obs/N`,
                       TRUE ~ -999)) %>%
  mutate(`P(C)` = ifelse(`P(C)` > 1, NA, `P(C)`)) %>%
  mutate(proportionOfEndowmentContributed = case_when(DV_behavior == "Contribution" & M != -999 & `Choice range upper` != -999 & `Choice range lower` != -999 ~ 
                                                        (M-as.numeric(`Choice range lower`))/(as.numeric(`Choice range upper`)-as.numeric(`Choice range lower`)),
                                                      DV_behavior == "Withdrawals" & M != -999 & `Choice range upper` != -999 ~ 
                                                        (as.numeric(`Choice range upper`) - M)/(as.numeric(`Choice range upper`)-as.numeric(`Choice range lower`)),
                                                      TRUE ~ -999)) %>%
  mutate(logPropContributed = case_when(`P(C)` != -999 ~ log(`P(C)`/(1 - `P(C)`)),
                                        proportionOfEndowmentContributed != -999 ~ log(proportionOfEndowmentContributed/(1 - proportionOfEndowmentContributed)),
                                        TRUE ~ -999),
         coefficientOfVariation = case_when(`P(C)` != -999 & n != -999 ~ 1 / (n * `P(C)`) + 1 / (n * (1 - `P(C)`)),
                                            DV_behavior == "Contribution" & proportionOfEndowmentContributed != -999 & SD != -999 ~ 
                                              (((SD^2) / ((M - as.numeric(`Choice range lower`))^2)) * 1 / (((1 - proportionOfEndowmentContributed)^2) * n)),
                                            DV_behavior == "Withdrawals" & proportionOfEndowmentContributed != -999 & SD != -999 ~ 
                                              (((SD^2) / ((as.numeric(`Choice range upper`) - M - as.numeric(`Choice range lower`))^2)) * 1 / (((1 - proportionOfEndowmentContributed)^2) * n)))) %>%
  select(study_ID:SD,proportionOfEndowmentContributed:coefficientOfVariation,everything())

# OBSERVATION-LEVEL DATA FRAME ################################################

# Create data frame with one row per observation, i.e., per possible
# combination of treatments within a variable. Treatments are denoted
# treatment_1 and treatment_2 within each observation, based on their
# treatment_ID.

# Separate variables with a single treatment from those with multiple
# treatments.

dataNomiss <- dataClean %>%
  mutate_at(vars(`Overall P(C)`:`P(E) contributed`, # Define NAs
                 `BS/WS`:SD),
            funs(na_if(., -999)))

# First set, variables with exactly one treatment.
# For these variables, the second treatment is empty.
dataOneTreatment <- dataNomiss %>%
  group_by(variable_ID) %>%
  filter(n() == 1) %>%
  ungroup() %>%
  select(variable_ID, treatment_ID) %>%
  dplyr::rename(treatment_1 = treatment_ID) %>%
  mutate(treatment_2 = NA)       # Set the second treatment to NA

# Second set, variables with more than one treatment
dataMultiTreatment <- dataNomiss %>%
  group_by(variable_ID) %>%
  filter(n() > 1)                # Exclude variable_IDs with a single treatment

# Get a data.table with every combination of treatment_IDs
# within each variable_ID
dataMultiTreatment <- setDT(dataMultiTreatment)[,
                                                {i1 <-  combn(treatment_ID, 2)
                                                list(i1[1,], i1[2,]) },
                                                by =  variable_ID]

# Merge dataMultiTreatment and dataOneTreatment
dataMultiTreatment <- dataMultiTreatment %>%
  dplyr::rename(treatment_1 = V1,
                treatment_2 = V2) %>%
  rbind(., dataOneTreatment)

# Merge the treatment indicators with data and generate unique effect_IDs
dataEffects <- merge(dataMultiTreatment, dataNomiss,
                     by.x = c("variable_ID", "treatment_1"),
                     by.y = c("variable_ID", "treatment_ID"),
                     all.x = TRUE) %>%
  merge(., data.frame(dataNomiss[,c("treatment_ID", "n", "P(C)", "M", "SD")]),
        by.x = "treatment_2", by.y = "treatment_ID",
        all.x = TRUE, suffixes = c(".t1", ".t2")) %>%
  rename(`P(C).t1` = `P(C)`,
         `P(C).t2` = `P.C.`) %>%
  mutate(effect_ID = ifelse(is.na(treatment_2), paste0(treatment_1, ".0"),
                            paste(treatment_1, sub("^.*\\.", "",
                                                   treatment_2), sep="."))) %>%
  select(study_ID, variable_ID, effect_ID,
         treatment_1, treatment_2,
         `Author(s) + year`:SD.t1,
         n.t2:SD.t2,everything())

# COMPUTING EFFECT SIZES ######################################################

# Effect size computation #####################################################

# Function to compute effect sizes.
# Takes a data.frame (specifically, dataNomiss)
# The specific effect size functions are hard-coded in effectSizeFunctions;
# right now, only es_from_means, es_from_prop, and es_from_rn are used.
# Output is a data.frame with effectSizeEstimate and effectSizeVar; also
# returned are effectSizeMeasure and effectSizeAlgorithm.
computeEffectSizes <- function(data){

  # Functions for effect-size computation for between-subjects designs are
  # hardcoded:
  effectSizeFunctions = list(es_from_means,
                             es_from_tn,
                             es_from_tdf,
                             es_from_prop,
                             es_from_rn,
                             es_from_means_dep)

  # computeEstimates iterates over the provided effectSizeFunctions
  # and applies them to the given data.frame and effect size measure.
  computeEstimates <- function(data, effsize) {
    effectSizes <- lapply(effectSizeFunctions, function(f) f(data, effsize)) %>%
      Reduce(function(x, y) merge(x, y, by="study", all=TRUE), .)

    # Name the columns.
    # Note that depending on the effect size measure, different numbers of
    # columns are generated.
    if(effsize == "r"){
      names(effectSizes) <- setdiff(c("study",
                                      paste(rep(c("means", "tn", "tdf", "prop", "rn", "means_dep"),
                                                each = 11),
                                            rep(c("es", "weight", "sample.size", "se",
                                                  "var", "ci.lo", "ci.hi", "fishersz",
                                                  "ci.lo.z", "ci.hi.z", "measure"),
                                                times = length(effectSizeFunctions)),
                                            sep = "_")),
                                    paste(rep(c("rn", "tn", "tdf", "means_dep"), each = 3),
                                          c("fishersz", "ci.lo.z", "ci.hi.z"), sep = "_"))
    }
    else{
      names(effectSizes) <- c("study",
                              paste(rep(c("means", "tn", "tdf", "prop", "rn", "means_dep"),
                                        each = 8),
                                    rep(c("es", "weight", "sample.size", "se",
                                          "var", "ci.lo", "ci.hi", "measure"),
                                        times = length(effectSizeFunctions)),
                                    sep = "_"))
    }
    return(effectSizes)
  }

  # Compute estimates of Cohen's d
  effectSizeD <- data %>%
    computeEstimates(data = ., effsize = "d") %>%
    dplyr::select(study,
                  ends_with("_es"),
                  ends_with("_var"),
                  ends_with("ci.lo"),
                  ends_with("ci.hi"),
                  ends_with("sample.size"),
                  ends_with("_measure")) %>%
    mutate(effect_ID = study,
           effectSizeEstimate = case_when(!is.na(means_dep_es) ~ means_dep_es,
                                          !is.na(means_es) ~ means_es,
                                          !is.na(tn_es) ~ tn_es,
                                          !is.na(tdf_es) ~ tdf_es,
                                          !is.na(prop_es) ~ prop_es,
                                          !is.na(rn_es) ~ rn_es),
           effectSizeVariance = case_when(!is.na(means_dep_es) ~ means_dep_var,
                                          !is.na(means_es) ~ means_var,
                                          !is.na(tn_es) ~ tn_var,
                                          !is.na(tdf_es) ~ tdf_var,
                                          !is.na(prop_es) ~ prop_var,
                                          !is.na(rn_es) ~ rn_var),
           effectSizeLowerLimit = case_when(!is.na(means_dep_es) ~ means_dep_ci.lo,
                                            !is.na(means_es) ~ means_ci.lo,
                                            !is.na(tn_es) ~ tn_ci.lo,
                                            !is.na(tdf_es) ~ tdf_ci.lo,
                                            !is.na(prop_es) ~ prop_ci.lo,
                                            !is.na(rn_es) ~ rn_ci.lo),
           effectSizeUpperLimit = case_when(!is.na(means_dep_es) ~ means_dep_ci.hi,
                                            !is.na(means_es) ~ means_ci.hi,
                                            !is.na(tn_es) ~ tn_ci.hi,
                                            !is.na(tdf_es) ~ tdf_ci.hi,
                                            !is.na(prop_es) ~ prop_ci.hi,
                                            !is.na(rn_es) ~ rn_ci.hi),
           effectSizeSampleSize = case_when(!is.na(means_dep_es) ~ means_dep_sample.size,
                                            !is.na(means_es) ~ means_sample.size,
                                            !is.na(tn_es) ~ tn_sample.size,
                                            !is.na(tdf_es) ~ tdf_sample.size,
                                            !is.na(prop_es) ~ prop_sample.size,
                                            !is.na(rn_es) ~ rn_sample.size),
           effectSizeMeasure = "d",
           effectSizeAlgorithm = case_when(!is.na(means_dep_es) ~ "means_dep",
                                           !is.na(means_es) ~ "means",
                                           !is.na(tn_es) ~ "tn",
                                           !is.na(tdf_es) ~ "tdf",
                                           !is.na(prop_es) ~ "prop",
                                           !is.na(rn_es) ~ "rn")) %>%
    select(effect_ID, effectSizeEstimate:effectSizeAlgorithm)

  # Compute estimates of r
  effectSizeR <- data %>%
    computeEstimates(data = ., effsize = "r") %>%
    dplyr::select(study,
                  ends_with("_es"),
                  ends_with("_var"),
                  ends_with("ci.lo"),
                  ends_with("ci.hi"),
                  ends_with("sample.size"),
                  ends_with("_measure")) %>%
    mutate(effect_ID = study,
           effectSizeEstimate = case_when(!is.na(rn_es) ~ rn_es,
                                          !is.na(means_es) ~ means_es,
                                          !is.na(tn_es) ~ tn_es,
                                          !is.na(tdf_es) ~ tdf_es,
                                          !is.na(prop_es) ~ prop_es),
           effectSizeVariance = case_when(!is.na(rn_es) ~ rn_var,
                                          !is.na(means_es) ~ means_var,
                                          !is.na(tn_es) ~ tn_var,
                                          !is.na(tdf_es) ~ tdf_var,
                                          !is.na(prop_es) ~ prop_var),
           effectSizeLowerLimit = case_when(!is.na(rn_es) ~ rn_ci.lo,
                                            !is.na(means_es) ~ means_ci.lo,
                                            !is.na(tn_es) ~ tn_ci.lo,
                                            !is.na(tdf_es) ~ tdf_ci.lo,
                                            !is.na(prop_es) ~ prop_ci.lo),
           effectSizeUpperLimit = case_when(!is.na(rn_es) ~ rn_ci.hi,
                                            !is.na(means_es) ~ means_ci.hi,
                                            !is.na(tn_es) ~ tn_ci.hi,
                                            !is.na(tdf_es) ~ tdf_ci.hi,
                                            !is.na(prop_es) ~ prop_ci.hi),
           effectSizeSampleSize = case_when(!is.na(rn_es) ~ rn_sample.size,
                                            !is.na(means_es) ~ means_sample.size,
                                            !is.na(tn_es) ~ tn_sample.size,
                                            !is.na(tdf_es) ~ tdf_sample.size,
                                            !is.na(prop_es) ~ prop_sample.size),
           effectSizeMeasure = "r",
           effectSizeAlgorithm = case_when(!is.na(rn_es) ~ "rn",
                                           !is.na(means_es) ~ "means",
                                           !is.na(tn_es) ~ "tn",
                                           !is.na(tdf_es) ~ "tdf",
                                           !is.na(prop_es) ~ "prop")) %>%
    select(effect_ID, effectSizeEstimate:effectSizeAlgorithm)

  # Combine estimates for d and r
  effectSizes <- rbind(effectSizeD, effectSizeR)
  
  # treatment_IDs with DV_behavior == "Withdrawals"
  withdrawals <- data %>% filter(DV_behavior == "Withdrawals")

  # Merge in treatment_IDs and DV specification.
  effectSizes <- data %>%
    dplyr::select(study_ID, effect_ID, treatment_1, treatment_2,
                  DV_behavior, DV_periods, DV_percentage) %>%
    merge(., effectSizes, by = "effect_ID", all.x = TRUE) %>%
    mutate(effect_ID = ifelse(is.na(effectSizeMeasure),
                              paste(effect_ID, "0", sep = "."),
                              paste(effect_ID, effectSizeMeasure, sep = "."))) %>%
    # Invert effect sizes from withdrawals
    mutate(esLL = effectSizeLowerLimit,
           esUL = effectSizeUpperLimit) %>%
    mutate(effectSizeEstimate = case_when(treatment_1 %in% withdrawals$treatment_1 ~ -effectSizeEstimate,
                                          TRUE ~ effectSizeEstimate),
           effectSizeLowerLimit = case_when(treatment_1 %in% withdrawals$treatment_1 ~ esUL,
                                            TRUE ~ esLL),
           effectSizeUpperLimit = case_when(treatment_1 %in% withdrawals$treatment_1 ~ esLL,
                                            TRUE ~ esUL)) %>%
    select(-esLL, -esUL)

  return(effectSizes)
}

# Compute effect sizes
effectSizes <- computeEffectSizes(dataEffects)

# Output files

write.csv(x=effectSizes, file=paste0(path,"/input/effect_sizes_computed.csv"))

write.csv(x=dataClean, file=paste0(path,"/input/data_clean.csv"), na="", quote=TRUE)
